### ISSUE TITLE
Differentiate between named functions and getter/setter functions.

### ISSUE THREAD

**Issue by Droogans** [eslint/eslint#5520]
This is not a new rule, but an extension of the core `space-before-function-paren` rule. I am filing this as part of my research efforts to answer [a stackoverflow question](http://stackoverflow.com/questions/35874314/space-before-function-paren-in-inline-getter-setter-functions) I made today.

**When does this rule warn? Please describe and show example code:**

This rule adds a third parameter to `space-before-function-paren` to allow for custom spacing for getter/setter functions.

```js
"space-before-function-paren": [2, {
    "anonymous": "always",
    "named": "always",
    "getterSetter": "never"
}]
```

This allows for code like this to stop throwing errors when using `"always"` for `"named"` functions.

```js
var myObject = {
    get myProperty() { return 'foo'; }
};
```

This aligns getter definitions with what [mdn recommends in their example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get#Syntax).

**Is this rule preventing an error or is it stylistic?**

Stylistic.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

This rule seeks to extend an existing ruleset. It makes sense to include it as an option in an already well-known ruleset, instead of isolating it in its own rule.

**Are you willing to create the rule yourself?**

If you're ok with me altering a core rule. I see that `get` and `set` are [already tracked under `parent.kind`](https://github.com/eslint/eslint/blob/288d3b27b80f492e5cf32c41a688bafcfc76f6c1/lib/rules/space-before-function-paren.js#L44-L46) when evaluating named functions explicitly. This means that the interpretation of how getter and setters are defined will need to be decoupled from named functions. If I'm able to make a clean break at the definition of what a named function is, it should be as simple as wrapping an if statement around that area of the code.

If I'm unable to make a breaking change that way, I may still be able to support some kind of a legacy definition of named functions as including getters/setters, and then over-ride that definition to include a finer grained definition if and only if they include the third option explicitly. However, just writing all that out made me cringe. I imagine it would look and feel much worse when implemented. Exposing variable definitions of core concepts of the ESLint parser sounds like a terrible precedent to set.

**Comment by ilyavolodin** [comment]
Could you provide code sample that would currently be flagged now, but you would like to be able to pass, and another that should fail with the new option? We also have a separate template for rule options requests here: http://eslint.org/docs/developer-guide/contributing/changes

**Comment by Droogans** [comment]
Including the outline from the "Bug" template here to condense the issues created.

**What version are you using?**

v2.2.0

**What did you do?**

I set my `space-before-function-paren as such:

```js
"space-before-function-paren": [2, { "anonymous": "always", "named": "always" }]
```

I then audited the various usage of functions in my project, and compared them to ESLints output.

> Defined functions: works fine.

```js
var myFunction = function (params) {
    return;
};
```

> Anonymous functions: works fine.

```js
return myFunction().then(function (result) {
    return;
});
```

> Hoisted functions: works fine.

```js
function myFunction (params) {
    return;
};
```

> Verbose getters/setters: works fine.

```js
Object.defineProperty(myObject, 'someProperty', {
    get: function () {
        return getSomething();
    },

    set: function (value) {
        doSomething(value);
    }
});
```
> Inline getters/setters: not working.

```js
var myObject = {
    get someProperty() { return getSomething(); }
    set someProperty(value) { doSomething(value); }
};
```

**What happened?**

I found that inline getters and setters were [considered the same things as named functions](https://github.com/eslint/eslint/blob/288d3b27b80f492e5cf32c41a688bafcfc76f6c1/lib/rules/space-before-function-paren.js#L44-L46).

**What did you expect to happen?**

I expected to be able to tell ESLint that I do not put spaces after my inline getter and setter function names and their parenthesis, but that I *do* put a space after my named function names and their parenthesis.

**Comment by BYK** [comment]
I think this can be an option.

**Comment by ilyavolodin** [comment]
@BYK Are you willing to champion inclusion of this option?

**Comment by BYK** [comment]
I'd rather get some more opinions from @eslint/eslint-team and other interested people. These are definitely functions but they are also methods. An option for methods would basically replicate all the options which is fine but may be people would prefer a separate rule. I'd wait a few days and if there are no objections, I'd go with the option path. Sounds good?

**Comment by nzakas** [comment]
If this is separating out accessors from named functions, it's a breaking change, so we need to be very certain that this is what we want.

Also, if we do want this, should we also have an option for concise methods?

**Comment by kaicataldo** [comment]
Closing this because we haven't made any progress in months. Thanks for contributing!
